Orange/data/table.py

```diff
@@ -33,146 +28,176 @@ def get_sample_datasets_dir():
-    __file__ = None
-    name = "untitled"
+    @property
+    def _constructor_sliced(self):
+        """Proper pandas extension as per http://pandas.pydata.org/pandas-docs/stable/internals.html"""
+        return TableSeries
+
+    @property
+    def _constructor_expanddim(self):
+        return TablePanel
+
+    @property
+    def domain(self):
+        # TODO: probably return a Domain object constructed from columns
+        return self.columns
+
+    @property
+    def columns_X(self):
+        """A read-only list of X column variables."""
+        return [c for c in self.columns if c in self._columns_X]
 
     @property
-    def columns(self):
+    def columns_Y(self):
+        """A read-only list of Y column variables."""
+        return [c for c in self.columns if c in self._columns_Y]
+
+    @property
+    def columns_meta(self):
+        """A read-only list of meta column variables."""
+        return [c for c in self.columns if c in self._columns_meta]
+
+    def _to_numpy(self, X=False, Y=False, meta=False, writable=False):
         """
-        A class whose attributes contain attribute descriptors for columns.
-        For a table `table`, setting `c = table.columns` will allow accessing
-        the table's variables with, for instance `c.gender`, `c.age` ets.
-        Spaces are replaced with underscores.
+        Exports a numpy matrix. The order is always X, Y, meta.
+        The columns are in the same order as in
+        Table.columns, Table.columns_X, Table.columns_Y, Table.columns_meta
+        If writable == False (default), the numpy writable flag is set to false.
+            This means write operations on this array will loudly fail. 
         """
-        return Columns(self.domain)
+        # TODO: only return numeric values here, need to transform
+        cols = []
+        cols += self.columns_X if X else []
+        cols += self.columns_Y if Y else []
+        cols += self.columns_meta if meta else []
+        res = self[cols].values
+        res.setflags(write=writable)
+        return res
 
-    _next_instance_id = 0
-    _next_instance_lock = Lock()
+    @property
+    def X(self):
+        """
+        Return a read-only numpy matrix of X.
+        Always a two-dimensional matrix, even for a single column.
+        The columns are in the same order as returned by Table.columns_X.
+        """
+        return self._to_numpy(X=True)
 
     @property
     def Y(self):
-        if self._Y.shape[1] == 1:
-            return self._Y[:, 0]
-        return self._Y
+        """
+        Return a read-only numpy array/matrix of Y.
+        If there is only one column, returns a one-dimensional array,
+            otherwise, returns a two-dimensional matrix whose columns
+            are in the same order as returned by Table.columns_Y.
+        """
+        res = self._to_numpy(Y=True)
+        # TODO: comment on whether this separation is okay please
+        return res.flatten() if len(self.columns_Y) else res
```

Comment:
If there is only one Y column, this fails because the tuple has only one element. Anyway, this is a remnant of previous modifications and does not need to be here any more.